parser/src/lexer.rs

```diff
@@ -520,814 +410,512 @@ where
                 }
             }
             '+' => {
-                let tok_start = self.get_pos();
-                self.next_char();
-                if let Some('=') = self.window[0] {
-                    self.next_char();
-                    let tok_end = self.get_pos();
-                    self.emit((Tok::PlusEqual, TextRange::new(tok_start, tok_end)));
+                if self.cursor.eat_char('=') {
+                    Tok::PlusEqual
                 } else {
-                    let tok_end = self.get_pos();
-                    self.emit((Tok::Plus, TextRange::new(tok_start, tok_end)));
+                    Tok::Plus
                 }
             }
             '*' => {
-                let tok_start = self.get_pos();
-                self.next_char();
-                match self.window[0] {
-                    Some('=') => {
-                        self.next_char();
-                        let tok_end = self.get_pos();
-                        self.emit((Tok::StarEqual, TextRange::new(tok_start, tok_end)));
-                    }
-                    Some('*') => {
-                        self.next_char();
-                        match self.window[0] {
-                            Some('=') => {
-                                self.next_char();
-                                let tok_end = self.get_pos();
-                                self.emit((
-                                    Tok::DoubleStarEqual,
-                                    TextRange::new(tok_start, tok_end),
-                                ));
-                            }
-                            _ => {
-                                let tok_end = self.get_pos();
-                                self.emit((Tok::DoubleStar, TextRange::new(tok_start, tok_end)));
-                            }
-                        }
-                    }
-                    _ => {
-                        let tok_end = self.get_pos();
-                        self.emit((Tok::Star, TextRange::new(tok_start, tok_end)));
+                if self.cursor.eat_char('=') {
+                    Tok::StarEqual
+                } else if self.cursor.eat_char('*') {
+                    if self.cursor.eat_char('=') {
+                        Tok::DoubleStarEqual
+                    } else {
+                        Tok::DoubleStar
                     }
+                } else {
+                    Tok::Star
                 }
             }
+
+            c @ ('%' | '!')
+                if self.mode == Mode::Jupyter
+                    && self.state.is_after_equal()
+                    && self.nesting == 0 =>
+            {
+                // SAFETY: Safe because `c` has been matched against one of the possible magic command prefix
+                self.lex_magic_command(MagicKind::try_from(c).unwrap())
+            }
+
+            c @ ('%' | '!' | '?' | '/' | ';' | ',')
+                if self.mode == Mode::Jupyter && self.state.is_new_logical_line() =>
```

Comment:
Am I correct to assume that `is_new_logical_line` will return `false` when inside a parenthesized statement? i.e., `(%foo)`

Reply:
Or, even
```
(
	%foo
)
```

Reply:
Yes. The *trick* is that the lexer never transitions into `State::AfterNewLine` or `State::NonEmptyLogicalLine` while inside parentheses (because everything inside parentheses forms the same logical line). 

https://github.com/astral-sh/RustPython-Parser/blob/46d940ff2bc7c0b03958e76d731bee587a2334fc/parser/src/lexer.rs#L849-L854

Okay, there's one situation where it could happen and this is when we insert a `Newline` at the end of the file. The Lexer does so regardless of whether all parentheses are closed or not. This should be fine because we already reached the end of the file.